server/models/UploadedFile.ts

```diff
@@ -0,0 +1,160 @@
+
+type ImageDataShape = {
+  width: number;
+  height: number;
+  blurHash: string;
+};
+
+// Constants
+export const SUPPORTED_FILE_TYPES_IMAGES = ['image/png', 'image/jpeg', 'image/gif'] as const;
+export const SUPPORTED_FILE_TYPES = [...SUPPORTED_FILE_TYPES_IMAGES, 'application/pdf'] as const;
+export const SUPPORTED_FILE_EXTENSIONS: Record<SUPPORTED_FILE_TYPES_UNION, string> = {
+  'image/png': '.png',
+  'image/jpeg': '.jpg',
+  'image/gif': '.gif',
+  'application/pdf': '.pdf',
+} as const;
+export const SUPPORTED_FILE_KINDS = [
+  // Base fields
+  'ACCOUNT_AVATAR',
+  'ACCOUNT_BANNER',
+  'EXPENSE_ATTACHED_FILE',
+  'EXPENSE_ITEM',
+  // Rich text fields
+  'ACCOUNT_LONG_DESCRIPTION',
+  'UPDATE',
+  'COMMENT',
+  'TIER_LONG_DESCRIPTION',
+  'ACCOUNT_CUSTOM_EMAIL',
+] as const;
+
+/**
+ * A file uploaded to our S3 bucket.
+ */
+class UploadedFile extends Model<InferAttributes<UploadedFile>, InferCreationAttributes<UploadedFile>> {
+  declare id: CreationOptional<number>;
+  declare kind: CreationOptional<(typeof SUPPORTED_FILE_KINDS)[number]>;
+  declare fileName: CreationOptional<string>;
+  declare fileSize: CreationOptional<number>; // In bytes
+  declare fileType: CreationOptional<SUPPORTED_FILE_TYPES_UNION>;
+  declare url: string;
+  declare data: CreationOptional<ImageDataShape>;
+  // Temporal fields
+  declare createdAt: CreationOptional<Date>;
+  declare updatedAt: CreationOptional<Date>;
+  declare deletedAt: CreationOptional<Date>;
+  // Relationships
+  declare CreatedByUserId: ForeignKey<User['id']>;
+
+  // Association methods
+  declare getCreatedByUser: BelongsToGetAssociationMixin<User>;
+
+  // ==== Static methods ====
+  static isOpenCollectiveS3BucketURL(url: string): boolean {
+    return new RegExp(`^https://${config.aws.s3.bucket}\\.s3[.-]us-west-1.amazonaws.com/\\w+`).test(url);
+  }
+}
+
+UploadedFile.init(
+  {
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: DataTypes.INTEGER,
+    },
+    kind: {
+      type: DataTypes.STRING,
+      allowNull: false,
+      validate: {
+        notNull: true,
+        notEmpty: true,
+        isIn: [SUPPORTED_FILE_KINDS],
+      },
+    },
+    fileName: {
+      type: DataTypes.STRING,
+      allowNull: true,
+    },
+    fileSize: {
+      type: DataTypes.INTEGER,
+      allowNull: true,
+      validate: {
+        isNumeric: true,
+        min: 0,
+        // No max here as its defined when uploading, this is only for reference
+      },
+    },
+    fileType: {
+      type: DataTypes.STRING,
+      allowNull: false,
+      validate: {
+        notNull: true,
+        notEmpty: true,
+        isIn: [SUPPORTED_FILE_TYPES],
+      },
+    },
+    url: {
+      type: DataTypes.STRING,
+      allowNull: false,
+      unique: true,
+      validate: {
+        notNull: true,
+        notEmpty: true,
+        isUrl: true,
+        isValidURL(url: string): void {
```

Comment:
I used minio to run this PR locally as a replacement for s3, would be nice to allow localhost URLs here in development
http://localhost:9000/opencollective-dvl/dd47d3c0-9c42-11ed-a0b2-b390eb1764c5